src/controller/cmcd-controller.ts

```diff
@@ -0,0 +1,536 @@
+      };
+    };
+  }
+
+  /**
+   * Create a playlist loader
+   *
+   * @returns {FragmentLoaderConstructor | undefined}
+   */
+  private createFragmentLoader(): FragmentLoaderConstructor | undefined {
+    const { fLoader } = this.config;
+    const apply = this.applyFragmentData;
+    const Ctor = fLoader || (this.config.loader as FragmentLoaderConstructor);
+
+    // @ts-ignore
+    return function (config: HlsConfig) {
+      const loader = new Ctor(config);
+      this.abort = () => loader.abort();
+      this.destroy = () => loader.destroy();
+      this.load = (
+        context: FragmentLoaderContext,
+        config: LoaderConfiguration,
+        callbacks: LoaderCallbacks<FragmentLoaderContext>
+      ) => {
+        apply(context);
+        loader.load(context, config, callbacks);
+      };
+      Object.defineProperty(this, 'stats', {
+        get: () => loader.stats,
+      });
+    };
+  }
+
+  /**
+   * Generate a random v4 UUI
+   *
+   * @returns {string}
+   */
+  static uuid(): string {
+    const url = URL.createObjectURL(new Blob());
+    const uuid = url.toString();
+    URL.revokeObjectURL(url);
+    return uuid.substr(uuid.lastIndexOf('/') + 1);
+  }
+
+  /**
+   * Serialize a CMCD data object according to the rules defined in the
+   * section 3.2 of
+   * [CTA-5004](https://cdn.cta.tech/cta/media/media/resources/standards/pdfs/cta-5004-final.pdf).
+   *
+   * @param {CMCD} data The CMCD data object
+   * @returns {string}
+   */
+  static serialize(data: CMCD): string {
+    const results: string[] = [];
+    const isValid = (value: any) =>
+      !Number.isNaN(value) && value != null && value !== '' && value !== false;
+    const toRounded = (value: number) => Math.round(value);
+    const toHundred = (value: number) => toRounded(value / 100) * 100;
+    const toUrlSafe = (value: string) => encodeURIComponent(value);
+    const formatters = {
+      br: toRounded,
+      d: toRounded,
+      bl: toHundred,
+      dl: toHundred,
+      mtp: toHundred,
+      nor: toUrlSafe,
+      rtp: toHundred,
+      tb: toRounded,
+    };
+
+    const keys = Object.keys(data || {}).sort();
+
+    for (const key of keys) {
+      let value = data[key];
+
+      // ignore invalid values
+      if (!isValid(value)) {
+        continue;
+      }
+
+      // Version should only be reported if not equal to 1.
+      if (key === 'v' && value === 1) {
+        continue;
+      }
+
+      // Playback rate should only be sent if not equal to 1.
+      if (key == 'pr' && value === 1) {
+        continue;
+      }
+
+      // Certain values require special formatting
+      const formatter = formatters[key];
+      if (formatter) {
+        value = formatter(value);
+      }
+
+      // Serialize the key/value pair
+      const type = typeof value;
+      let result: string;
+
+      if (type === 'string' && key !== 'ot' && key !== 'sf' && key !== 'st') {
+        result = `${key}="${value.replace(/"/g, '\\"')}"`;
```

Comment:
>7. Any value of type String MUST be enclosed by opening and closing double
quotes Unicode 0x22. Double quotes and backslashes MUST be escaped using a
backslash "\" Unicode 0x5C character. Any value of type Token does not require
quoting. 

This implies we should also be escaping backslash?

```suggestion
        result = `${key}="${value.replace(/\\/g, '\\\\').replace(/"/g, '\\"')"`;
```
?

Also I think we should do this for the `symbol` type and the final `else`? I'm wondering why this excludes certain keys above?

Reply:
The excluded keys have "token" values according to the spec, and should not be quoted.

Reply:
Ah right. Not quoted but probably should still be escaped?

Reply:
They don't need to be escaped because they are single character values. See: 
https://github.com/video-dev/hls.js/blob/932c58a3f1bb14ffe62411b6ef4f1de7fadafbc4/src/types/cmcd.ts#L9-L37

Reply:
The code has been updated to account for `\` and `"`.

Reply:
Great thanks

Please could we invert it then so that we have a case that handles the token types (ot, sf etc) and then another that handles the type for all other strings, and does the escape? I think it’s slightly more readable because it keeps the token logic together instead of it being split in the if and else block further down

Also if there are enum values could we import and use the enum instead?

Finally to be on the safe side I think we should json stringify the symbol description too? Are we actually expecting symbols?

Reply:
The spec allows for custom fields to be pass along side the standard CMCD fields. Symbols are the way to tell the serializer that the value represents a custom token, and tokens are not quoted or escaped.

Reply:
The token processing has been moved to its own if block.